In [1]:
pip install efficient-apriori

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [3]:
#import package
from efficient_apriori import apriori

In [5]:
#buat data transaksi
transactions = [('Bread','Milk'),
               ('Bread', 'Diaper', 'Beer', 'Eggs'),
               ('Milk', 'Diaper', 'Beer', 'Coke'),
               ('Bread', 'Milk', 'Diaper', 'Beer'),
               ('Bread', 'Milk', 'Diaper', 'Coke')]

Pada data di atas terdiri dari 5 row data, dimana setiap row menunjukan satu transaksi pada suatu
toko. Sehingga dari data tersebut kita memiliki 5 transaksi.


In [8]:
#terapkan algoritma Apriori
itemsets, rules = apriori(transactions, min_support= 0.4, min_confidence= 0.6)

Dimana kita set minimum support adalah 0.4 dan minimum confidence adalah 0.6. Setelah itu kita
coba tampilkan itemsetsnya dengan ketikkan kode di bawah ini.

In [9]:
for j in range(1, len(itemsets)+1):
  print(itemsets[j])
  print("\n")

{('Bread',): 4, ('Milk',): 4, ('Diaper',): 4, ('Beer',): 3, ('Coke',): 2}


{('Beer', 'Bread'): 2, ('Beer', 'Diaper'): 3, ('Beer', 'Milk'): 2, ('Bread', 'Diaper'): 3, ('Bread', 'Milk'): 3, ('Coke', 'Diaper'): 2, ('Coke', 'Milk'): 2, ('Diaper', 'Milk'): 3}


{('Beer', 'Bread', 'Diaper'): 2, ('Beer', 'Diaper', 'Milk'): 2, ('Bread', 'Diaper', 'Milk'): 2, ('Coke', 'Diaper', 'Milk'): 2}




Selanjutnya jika kita ingin menampilkan association rules yang terbentuk kita bisa ketikkan kode
di bawah ini.

In [10]:
for i in range(len(rules)):
  print(rules[i])
  print()

{Beer} -> {Bread} (conf: 0.667, supp: 0.400, lift: 0.833, conv: 0.600)

{Diaper} -> {Beer} (conf: 0.750, supp: 0.600, lift: 1.250, conv: 1.600)

{Beer} -> {Diaper} (conf: 1.000, supp: 0.600, lift: 1.250, conv: 200000000.000)

{Beer} -> {Milk} (conf: 0.667, supp: 0.400, lift: 0.833, conv: 0.600)

{Diaper} -> {Bread} (conf: 0.750, supp: 0.600, lift: 0.938, conv: 0.800)

{Bread} -> {Diaper} (conf: 0.750, supp: 0.600, lift: 0.938, conv: 0.800)

{Milk} -> {Bread} (conf: 0.750, supp: 0.600, lift: 0.938, conv: 0.800)

{Bread} -> {Milk} (conf: 0.750, supp: 0.600, lift: 0.938, conv: 0.800)

{Coke} -> {Diaper} (conf: 1.000, supp: 0.400, lift: 1.250, conv: 200000000.000)

{Coke} -> {Milk} (conf: 1.000, supp: 0.400, lift: 1.250, conv: 200000000.000)

{Milk} -> {Diaper} (conf: 0.750, supp: 0.600, lift: 0.938, conv: 0.800)

{Diaper} -> {Milk} (conf: 0.750, supp: 0.600, lift: 0.938, conv: 0.800)

{Bread, Diaper} -> {Beer} (conf: 0.667, supp: 0.400, lift: 1.111, conv: 1.200)

{Beer, Diaper} -> {Bread}

**Contoh kedua dengan menggunakan data dari csv file:**

In [11]:
import pandas as pd
from math import isnan
from efficient_apriori import apriori

In [12]:
from google.colab import drive
drive.mount ('/content/drive')

Mounted at /content/drive


In [38]:
# import data
store_data = pd.read_csv ("/content/drive/MyDrive/Dataset/contoh_data.csv", header=None)
store_data.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [39]:
#cek tipe data
store_data.dtypes

0     object
1     object
2     object
3     object
4     object
5     object
6     object
7     object
8     object
9     object
10    object
11    object
12    object
13    object
14    object
15    object
16    object
17    object
18    object
19    object
dtype: object

In [40]:
# Setelah itu kita ubah data yang bertipe dataframe ke tipe data tuple. Karena untuk library apriori
# hanya dapat menggunakan data yang bertipe tuple.
store_data = [tuple(row) for row in store_data.values.tolist()]
print(store_data)
print("\n")

[('shrimp', 'almonds', 'avocado', 'vegetables mix', 'green grapes', 'whole weat flour', 'yams', 'cottage cheese', 'energy drink', 'tomato juice', 'low fat yogurt', 'green tea', 'honey', 'salad', 'mineral water', 'salmon', 'antioxydant juice', 'frozen smoothie', 'spinach', 'olive oil'), ('burgers', 'meatballs', 'eggs', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan), ('chutney', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan), ('turkey', 'avocado', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan), ('mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan), ('low fat yogurt', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan), ('whole wheat pasta', 'french fries', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, n

In [41]:
# lakukan praproses data dengan menghilangkan data yang bertipe float dan jika data tersebut NaN values.
for y in range(0,len(store_data)):
  store_data[y] = [x for x in store_data[y] if not (
      type(x) == float #drop semua nilai float
      and isnan(x) 
  )]
print(store_data)
print("\n")

[['shrimp', 'almonds', 'avocado', 'vegetables mix', 'green grapes', 'whole weat flour', 'yams', 'cottage cheese', 'energy drink', 'tomato juice', 'low fat yogurt', 'green tea', 'honey', 'salad', 'mineral water', 'salmon', 'antioxydant juice', 'frozen smoothie', 'spinach', 'olive oil'], ['burgers', 'meatballs', 'eggs'], ['chutney'], ['turkey', 'avocado'], ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'], ['low fat yogurt'], ['whole wheat pasta', 'french fries'], ['soup', 'light cream', 'shallot'], ['frozen vegetables', 'spaghetti', 'green tea'], ['french fries']]




Selajutnya kita cari association rulesnya dari data tersebut, seperti pada kode dibawah

In [42]:
#terapkan algoritma Apriori
itemsets, rules = apriori(store_data, min_support= 0.2, min_confidence= 0.5)

for i in range(len(rules)):
  print(rules[i])
  print()

{mineral water} -> {green tea} (conf: 1.000, supp: 0.200, lift: 3.333, conv: 700000000.000)

{green tea} -> {mineral water} (conf: 0.667, supp: 0.200, lift: 3.333, conv: 2.400)



**Analisa Hasil:**

Dari hasil association rule untuk data transaksi yang ada, dapat disimpulkan bahwa pembelian
mineral water dan grean tea secara bersamaan memiliki support 0.2 yang artinya terdapat 20%
penjualan dari seuruh transaksi. Sedangkan berdasarkan confidence setiap pembelian mineral
water maka 100% akan membeli grean tea karena confidence = 1.0. Namun setiap pembelian
green tea maka ada kemungkinan 66.7% akan membeli mineral water karena confidence 0.667.